Code/GraphMol/Abbreviations/Abbreviations.cpp

```diff
@@ -0,0 +1,256 @@
+      connectingAtom->setProp(RDKit::common_properties::_displayLabelW,
+                              amatch.abbrev.displayLabelW);
+    }
+
+    connectingAtom->setFormalCharge(0);
+    connectingAtom->setAtomicNum(0);
+    connectingAtom->setIsotope(0);
+    connectingAtom->setIsAromatic(false);
+
+    // set the hybridization so these are drawn linearly
+    connectingAtom->setHybridization(Atom::HybridizationType::SP);
+
+    for (unsigned int i = 2; i < amatch.match.size(); ++i) {
+      const auto& pr = amatch.match[i];
+      CHECK_INVARIANT(!toRemove[pr.second], "overlapping matches");
+      toRemove.set(pr.second);
+      // if there's a molecule associated with the match, check to see if
+      // additional bonds need to be formed
+      if (amatch.abbrev.mol &&
+          mol.getAtomWithIdx(pr.second)->getDegree() >
+              amatch.abbrev.mol->getAtomWithIdx(pr.first)->getDegree()) {
+        for (const auto& nbri : boost::make_iterator_range(
+                 mol.getAtomNeighbors(mol.getAtomWithIdx(pr.second)))) {
+          const auto& nbr = mol[nbri];
+          auto nbrIdx = nbr->getIdx();
+          // if this neighbor isn't in the match:
+          if (!std::any_of(amatch.match.begin(), amatch.match.end(),
+                           [&](const std::pair<int, int>& tpr) {
+                             return tpr.second == rdcast<int>(nbrIdx);
+                           })) {
+            mol.addBond(nbrIdx, connectIdx, Bond::BondType::SINGLE);
+          }
+        }
+      }
+    }
+    // make connections between any extraAttachAtoms and the connection point
+    for (auto oaidx : amatch.abbrev.extraAttachAtoms) {
+      mol.addBond(oaidx, connectIdx, Bond::BondType::SINGLE);
+    }
+  }
+  for (unsigned int i = toRemove.size(); i > 0; --i) {
+    if (toRemove[i - 1]) {
+      mol.removeAtom(i - 1);
+    }
+  }
+}
+
+void labelMatches(RWMol& mol, const std::vector<AbbreviationMatch>& matches) {
+  for (const auto& amatch : matches) {
+    // throughout this remember that atom 0 in the match is the dummy
+    SubstanceGroup sg(&mol, "SUP");
+    sg.setProp("LABEL", amatch.abbrev.label);
+
+    for (unsigned int i = 1; i < amatch.match.size(); ++i) {
+      const auto& pr = amatch.match[i];
+      sg.addAtomWithIdx(pr.second);
+    }
+    auto bnd =
+        mol.getBondBetweenAtoms(amatch.match[0].second, amatch.match[1].second);
+    CHECK_INVARIANT(bnd, "bond to attachment point not found");
+    sg.addBondWithIdx(bnd->getIdx());
+    sg.addAttachPoint(amatch.match[1].second, amatch.match[0].second, "1");
+    addSubstanceGroup(mol, sg);
+  }
+}
+
+std::vector<AbbreviationMatch> findApplicableAbbreviationMatches(
+    const ROMol& mol, const std::vector<AbbreviationDefinition>& abbrevs,
+    double maxCoverage) {
+  std::vector<AbbreviationMatch> res;
+  auto nAtoms = mol.getNumAtoms();
+  if (!nAtoms || abbrevs.empty()) {
+    return res;
+  }
+
+  MolOps::fastFindRings(mol);
```

Comment:
Haven't we had issues with fast find rings having issues down stream?  If we call fast find rings, should we remove the rings at the end?

Reply:
(This might be on a copy)

Reply:
It's not on a copy, but FastFindRings() will only do something on molecules that don't already have ring info (i.e. unsanitized molecules), so I don't think it's a problem.

Reply:
For downstream effects, I was referring to things like this:

https://github.com/rdkit/rdkit/issues/1021

Where GetSymmSSSR is what the sanitizer would call instead.

Reply:
Yeah, I appreciate the problem. One should really sanitize things before using them, but people won’t. 
I’ll close this

Reply:
Yeah, I appreciate the problem. One should really sanitize things before using them, but people won’t. 
I’ll close this